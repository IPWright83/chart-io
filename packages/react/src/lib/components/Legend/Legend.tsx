import React from "react";
import { useSelector } from "react-redux";

import { chartSelectors, IState } from "../../store";

const Legend = () => {
    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));

    // Only needed if xs or ys defined
    // chartReducer #29 - need to group scales together
    // so they can be retrieved from the selectors as a group
    // and determine which is the "series" fields if there are multiple series

    return (
        <foreignObject width={width} height={height} style={styles.foreignObject}>
            <div>test</div>
        </foreignObject>
    );
};

const styles = {
    canvas: {
        position: "absolute" as const,
    },
    foreignObject: {
        pointerEvents: "none" as const,
    },
};

export { Legend };
