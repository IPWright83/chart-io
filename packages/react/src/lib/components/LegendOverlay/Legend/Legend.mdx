import { Story, Canvas, Meta } from "@storybook/blocks";
import LinkTo from "@storybook/addon-links/react";

<Meta title="Components/LegendOverlay/Legend" />

# Legend Component

The `<Legend>` component allows us to create a legend containing multiple entries, which is useful for showing the information about the series forming a chart.

<Canvas>
    <Story id="components-legendoverlay-legend--default" />
</Canvas>

The legend itself displays rows of <LinkTo kind="components-legendoverlay-legend--scatter">`<LegendItem>`</LinkTo> components which offer their own configuration.

### `<Legend>` Props

> **Note:** Normally you don't interact with the Legend component directly, instead letting the <LinkTo kind="components-legendoverlay--default">`<LegendOverlay>`</LinkTo> component handle the displaying of the tooltip which provides all the required props.

| Prop            | Type     | Default   | Note                                                                                                          |
| --------------- | -------- | --------- | ------------------------------------------------------------------------------------------------------------- |
| `items`         | `array`  | undefined | The list if items that should be renders as `<LegendItem>`s                                                   |
| `positionStyle` | `object` | {}        | Additional css to control the position, typically containing just the top, left, bottom, right css properties |
| `formatters`    | `object` | {}        | A keyed collection of format functions `(name: string) => string`                                             |

<br />
