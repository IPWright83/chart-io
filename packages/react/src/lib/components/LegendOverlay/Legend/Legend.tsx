import type { ILegendFormatter, ILegendItem } from "@d3-chart/types";
import React from "react";
import { useSelector } from "react-redux";

import { LegendItem } from "./LegendItem";
import { chartSelectors, IState } from "../../../store";

export interface ILegendProps {
    /**
     * A style object controling the position of the toolip
     * @type {Object}
     */
    positionStyle?: Record<string, any>;
    /**
     * An array of legend items to display
     * @type {Array<LegendItem>}
     */
    items: Array<ILegendItem>;
    /**
     * A set of custom formatters for the Legend
     */
    formatters?: Record<string, ILegendFormatter>;
}

/**
 * Represents a Legend
 * @return The Legend component
 */
export function Legend({ items, positionStyle, formatters = {} }: ILegendProps) {
    const theme = useSelector((s: IState) => chartSelectors.theme(s));
    console.log(positionStyle);
    const style = {
        border: `thin solid ${theme.legend.border}`,
        display: "inline-block",
        ...positionStyle,
        padding: theme.legend.padding,
        background: theme.legend.background.toString(),
        opacity: theme.legend.opacity,
    };

    if (!items || items.length === 0) {
        return null;
    }

    return (
        <div className="d3-chart legend" style={style}>
            {items.map((item) => {
                /**
                 * A format is of the shape:
                 * {
                 *     formatFunc: (name: string) => string;
                 * }
                 */
                const formatter = formatters[item.name] || undefined;

                return <LegendItem key={`${item.name}`} format={formatter} {...item} />;
            })}
        </div>
    );
}
