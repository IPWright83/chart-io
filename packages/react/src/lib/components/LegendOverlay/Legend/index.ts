export * from "./Legend";
export * from "./LegendItem";
