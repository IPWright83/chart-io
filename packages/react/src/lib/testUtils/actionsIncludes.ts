/**
 * Assert that all the expected dispatch calls to the Redux store were seen
 * @param {string[]} dispatchCalls The actual calls
 * @param {string[]} expectedCalls The expected calls
 */
export function actionsIncludes(dispatchCalls: string[], expectedCalls: string[]) {
    expectedCalls.forEach((expected) => {
        expect(dispatchCalls).toContain(expected);
        dispatchCalls.splice(dispatchCalls.indexOf(expected), 1);
    });
}
