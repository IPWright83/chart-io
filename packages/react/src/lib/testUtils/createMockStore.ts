import { defaultChartState, defaultEventState, IStore } from "../store";

/**
 * Creates a mock Redux store for testing with
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStore(state: { chart?: any; event?: any }): IStore {
    return {
        getState: () => ({
            chart: {
                ...{ ...defaultChartState, id: "test" },
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        dispatch: jest.fn(),
        subscribe: jest.fn(),
    } as unknown as IStore;
}

/**
 * Creates a mock Redux store for Storybook
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStorybookStore(state: { chart?: any; event?: any }) {
    return {
        getState: () => ({
            chart: {
                ...defaultChartState,
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        dispatch: () => {},
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        subscribe: () => {},
    } as unknown as IStore;
}
