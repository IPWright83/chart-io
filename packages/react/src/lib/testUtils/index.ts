export * from "./actionsIncludes";
export * from "./FakeMouseEvent";
export * from "./createMockStore";
export * from "./getBuffer";
export * from "./wait";
export * from "./renderChart";
export * from "./testEvents";
