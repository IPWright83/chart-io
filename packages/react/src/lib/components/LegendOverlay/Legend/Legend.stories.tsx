import React from "react";
import { Provider } from "react-redux";

import { Legend } from "./Legend";

import { createMockStorybookStore } from "../../../testUtils";

export default {
  title: "Components/LegendOverlay/Legend",
  component: Legend,
  parameters: {
    docs: {
      transformSource: (src) => {
        src = src.replace(/data={\[.*?\]}/gs, "data={[ ...dataset ]}");
        src = src.replaceAll(/undefined,?/g, "");
        src = src.replace(/^\s*\n/gm, "");
        return src;
      },
    },
    chromatic: { delay: 300 },
  },
};

const LegendTemplate = (args) => {
  const store = createMockStorybookStore({});

  return (
    <Provider store={store}>
      <Legend items={args.items} formatters={args.formatters} />
    </Provider>
  );
};

export const Default = {
  name: "All Series Types",
  render: LegendTemplate,
  args: {
    items: [
      {
        name: "Line Series with a very long title that should be truncated at some point",
        icon: "line",
        color: "steelblue",
      },
      { name: "Scatter Series", icon: "circle", color: "steelblue" },
      { name: "Bar Series", icon: "square", color: "steelblue" },
      { name: "Column Series", icon: "square", color: "steelblue" },
      { name: "Area Series", icon: "square", color: "steelblue" },
      { name: "Timestamp", color: "steeblue", icon: "none" },
    ],
  },
};

export const CustomFormatters = {
  name: "Custom Formatting",
  render: LegendTemplate,
  args: {
    items: [{ name: "x", icon: "none" }],
    formatters: {
      x: (name) => `~~~${name}~~~`,
    },
  },
};
