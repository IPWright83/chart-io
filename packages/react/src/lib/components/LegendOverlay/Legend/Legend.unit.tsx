import type { IColor } from "@d3-chart/types";
import { Provider } from "react-redux";
import React from "react";
import { render } from "@testing-library/react";

import { createMockStore } from "../../../testUtils";

import { Legend } from "./Legend";

describe("Legend", () => {
    const store = createMockStore({});

    it("should render correctly", async () => {
        const positionStyle = {
            position: "absolute",
            left: 10,
            right: 20,
            top: 30,
            bottom: 40,
        };

        const items = [
            { name: "a", icon: "circle" as const, fill: "blue" as IColor },
            { name: "b", icon: "square" as const, fill: "orange" as IColor },
            { name: "c", icon: "line" as const, fill: "green" as IColor },
        ];

        const { asFragment } = render(
            <Provider store={store}>
                <Legend items={items} positionStyle={positionStyle} />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render nothing if there are no items", async () => {
        const { asFragment } = render(
            <Provider store={store}>
                <Legend items={[]} />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
