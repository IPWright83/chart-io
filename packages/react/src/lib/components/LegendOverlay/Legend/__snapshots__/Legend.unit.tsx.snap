// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend should render correctly 1`] = `
<DocumentFragment>
  <div
    class="d3-chart legend"
    style="border: thin solid #ccc; display: flex; flex-direction: column; flex-wrap: wrap; position: absolute; left: 10px; right: 20px; top: 30px; bottom: 40px; padding: 8px; background: rgb(255, 255, 255); opacity: 1; overflow: hidden;"
  >
    <div
      class="d3-chart legend-item"
      style="display: flex; flex-direction: row; user-select: none; pointer-events: none; color: rgb(0, 0, 0); font-size: 12px; font-family: sans-serif;"
    >
      <div
        class="d3-chart indicator-circle"
        style="background: blue; width: 10px; height: 10px; margin-top: 4px; margin-left: 5px; margin-right: 5px; border-radius: 50%; flex-shrink: 0;"
      />
      <div
        class="d3-chart legend-values"
        style="display: flex; flex-direction: row; justify-content: space-between; width: 100%; flex-grow: 1; flex-shrink: 1;"
      >
        <span
          class="d3-chart legend-series-name"
          style="margin-right: 15px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; flex-grow: 1; flex-shrink: 1;"
        >
          a
        </span>
      </div>
    </div>
    <div
      class="d3-chart legend-item"
      style="display: flex; flex-direction: row; user-select: none; pointer-events: none; color: rgb(0, 0, 0); font-size: 12px; font-family: sans-serif;"
    >
      <div
        class="d3-chart indicator-square"
        style="background: orange; height: 10px; width: 10px; margin-top: 4px; margin-right: 5px; margin-left: 5px; flex-shrink: 0;"
      />
      <div
        class="d3-chart legend-values"
        style="display: flex; flex-direction: row; justify-content: space-between; width: 100%; flex-grow: 1; flex-shrink: 1;"
      >
        <span
          class="d3-chart legend-series-name"
          style="margin-right: 15px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; flex-grow: 1; flex-shrink: 1;"
        >
          b
        </span>
      </div>
    </div>
    <div
      class="d3-chart legend-item"
      style="display: flex; flex-direction: row; user-select: none; pointer-events: none; color: rgb(0, 0, 0); font-size: 12px; font-family: sans-serif;"
    >
      <div
        class="d3-chart indicator-line"
        style="background: green; width: 10px; height: 3px; margin-top: 7px; margin-left: 5px; margin-right: 5px; flex-shrink: 0;"
      />
      <div
        class="d3-chart legend-values"
        style="display: flex; flex-direction: row; justify-content: space-between; width: 100%; flex-grow: 1; flex-shrink: 1;"
      >
        <span
          class="d3-chart legend-series-name"
          style="margin-right: 15px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; flex-grow: 1; flex-shrink: 1;"
        >
          c
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Legend should render nothing if there are no items 1`] = `<DocumentFragment />`;
